CMakeLists.txt

```diff
@@ -1277,3 +1277,23 @@ if(BUILD_BUNDLE_PTXAS AND USE_CUDA)
    endif()
    install(PROGRAMS "${PROJECT_BINARY_DIR}/ptxas" DESTINATION "${CMAKE_INSTALL_BINDIR}")
 endif()
+
+# Opentelemetry Options: Required for including OpenTelemetry API headers for PyTorch
+# Let's give the same CXX that PyTorch uses to OpenTelemetry
+set(WITH_STL CXX${CMAKE_CXX_STANDARD})
+# We only want to take a dependency on the API headers
+set(WITH_API_ONLY ON)
+# Turn off building tests. OpenTelemetry Tests requires GTest and we don't need these for API
+set(BUILD_TESTING OFF)
+# Use Version 1. From documentation, version 1 is stable, version 2 is preview/experimental
+set(OPENTELEMETRY_ABI_VERSION_NO "1")
+# Install the API package config - this makes opentelemetry_api pkg available to binaries that need it
+set(OPENTELEMETRY_INSTALL "ON")
+include_directories(${CMAKE_CURRENT_SOURCE_DIR}/third_party/opentelemetry-cpp/api/include)
+# Set PROJECT_SOURCE_DIR so we can properly build opentelemetry-api - it needs
+# this variable set relative to opentelemetry install
+set(PROJECT_SOURCE_DIR "${CMAKE_CURRENT_SOURCE_DIR}/third_party/opentelemetry-cpp")
+add_subdirectory(third_party/opentelemetry-cpp/api)
```

Comment:
I'm a bit surprised about this, because OpenTelemetry claims that its API library is header only, so it seems to me that we can just add its header to include_directories and we're done, we don't need the rest of its cmake. Is there something else going on here?

Reply:
Here is an example of how we interact with pybind11, another header only library

```
# ---[ pybind11
if(USE_SYSTEM_PYBIND11)
  find_package(pybind11 CONFIG)
  if(NOT pybind11_FOUND)
    find_package(pybind11)
  endif()
  if(NOT pybind11_FOUND)
    message(FATAL "Cannot find system pybind11")
  endif()
else()
    message(STATUS "Using third_party/pybind11.")
    set(pybind11_INCLUDE_DIRS ${CMAKE_CURRENT_LIST_DIR}/../third_party/pybind11/include)
    install(DIRECTORY ${pybind11_INCLUDE_DIRS}
            DESTINATION ${CMAKE_INSTALL_PREFIX}
            FILES_MATCHING PATTERN "*.h")
endif()
message(STATUS "pybind11 include dirs: " "${pybind11_INCLUDE_DIRS}")
add_library(pybind::pybind11 INTERFACE IMPORTED)
target_include_directories(pybind::pybind11 SYSTEM INTERFACE ${pybind11_INCLUDE_DIRS})
target_link_libraries(pybind::pybind11 INTERFACE python::python)
if(APPLE)
  target_link_options(pybind::pybind11 INTERFACE -undefined dynamic_lookup)
endif()
```

Ignore the USE_SYSTEM_PYBIND11 branch, it's typically not used outside of distro packagers of PyTorch.

Reply:
So, I tried adding a random OTEL header inside a random C++ file - just for kicks as follows. My assumption was that users of OTEL would do things similar to what I'm doing.

```
diff --git a/torch/csrc/serialization.cpp b/torch/csrc/serialization.cpp
index 5a6cab19835..40cdb28ec97 100644
--- a/torch/csrc/serialization.cpp
+++ b/torch/csrc/serialization.cpp
@@ -5,6 +5,11 @@
 #include <c10/core/CPUAllocator.h>
 #include <torch/csrc/THP.h>
 #include <torch/csrc/serialization.h>
+#include <opentelemetry/common/timestamp.h>
+#include <opentelemetry/trace/tracer.h>
+
+using opentelemetry::common::SystemTimestamp;
+

 template <class io>
 Py_ssize_t doPartialRead(io fildes, void* buf, size_t nbytes);
@@ -27,6 +32,8 @@ static Py_ssize_t doPartialPythonWrite(

 template <>
 Py_ssize_t doPartialRead<int>(int fildes, void* buf, size_t nbytes) {
+  opentelemetry::trace::StartSpanOptions start;
+  start.start_system_time = SystemTimestamp(std::chrono::nanoseconds(300));
   return read(fildes, buf, nbytes);
 }

(/home/cpio/local/a/pytorch-env) [cpio@devvm17556.vll0 ~/local/pytorch (gh/c-p-i-o/otel)]$
```
When I tried to compile `torch`, the link step would fail unless I added all the mentioned `target_link`. But maybe I'm missing something?

Reply:
Yup! You're right. We don't need these if we are just using the header files.
I've removed these from the main file. 
We just need a single link in `torch/CMakeLists.txt` to allow it to bring in header files.
I tested manually by adding some code - this won't be checked in - as follows:
```
diff --git a/torch/csrc/serialization.cpp b/torch/csrc/serialization.cpp
index 5a6cab19835..40cdb28ec97 100644
--- a/torch/csrc/serialization.cpp
+++ b/torch/csrc/serialization.cpp
@@ -5,6 +5,11 @@
 #include <c10/core/CPUAllocator.h>
 #include <torch/csrc/THP.h>
 #include <torch/csrc/serialization.h>
+#include <opentelemetry/common/timestamp.h>
+#include <opentelemetry/trace/tracer.h>
+
+using opentelemetry::common::SystemTimestamp;
+

 template <class io>
 Py_ssize_t doPartialRead(io fildes, void* buf, size_t nbytes);
@@ -27,6 +32,8 @@ static Py_ssize_t doPartialPythonWrite(

 template <>
 Py_ssize_t doPartialRead<int>(int fildes, void* buf, size_t nbytes) {
+  opentelemetry::trace::StartSpanOptions start;
+  start.start_system_time = SystemTimestamp(std::chrono::nanoseconds(300));
   return read(fildes, buf, nbytes);
 }

(/home/cpio/local/a/pytorch-env) [cpio@devvm17556.vll0 ~/local/pytorch (gh/c-p-i-o/otel)]$
```
And now, compile works:
```
(/home/cpio/local/a/pytorch-env) [cpio@devvm17556.vll0 ~/local/pytorch (gh/c-p-i-o/otel)]$ MAX_JOBS=400 USE_CUDA=1 USE_MPI=0 BUILD_CAFFE2_OPS=0 USE_XNNPACK=0 USE_FBGEMM=0 USE_MKLDNN=0 USE_QNNPACK=0 BUILD_TEST=0 USE_GOLD_LINKER=1 USE_OPENMP=0 USE_PYTORCH_QNNPACK=0 USE_NNPACK=0 WITH_STL=ON WITH_ABI_VERSION_1=ON CMAKE_PREFIX_PATH="$CONDA_PREFIX" BLAS=MKL python3 setup.py develop
...

Using /data/users/cpio/a/pytorch-env/lib/python3.10/site-packages/mpmath-1.3.0-py3.10.egg
Finished processing dependencies for torch==2.4.0a0+git9ca5835
(/home/cpio/local/a/pytorch-env) [cpio@devvm17556.vll0 ~/local/pytorch (gh/c-p-i-o/otel)]$
```